src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,158 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.NoSuchElementException;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node head;
+    private Node tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        link(tail, value, null);
     }
 
     @Override
     public void add(T value, int index) {
+        if (size == index) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        Node foundNode = getElementByIndex(index);
+        link(foundNode.prev, value, foundNode);
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node foundNode = getElementByIndex(index);
+        return foundNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node foundNode = getElementByIndex(index);
+        T oldValue = foundNode.value;
+        foundNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node foundNode = getElementByIndex(index);
+        T value = foundNode.value;
+        unlink(foundNode);
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node foundNode = getElementByValue(object);
```

Comment:
```suggestion
        Node<T> current = getElementByValue(object);
```
Please, check all places where you should use generic type